### DISCUSSION TITLE
SVG Icons

### DISCUSSION THREAD

**Discussion by kentcdodds** [epicweb-dev/epic-stack#225] [Ideas]
I know how I like to do this now ([using SVG sprites](https://benadam.me/thoughts/react-svg-sprites/)). I need to set up a nice way to do this (maybe using @kiliman's [rmx-cli](https://github.com/kiliman/rmx-cli)?) so it's easy to add SVGs, but I'm posting this here so folks know I'm thinking about it!

**Comment by kiliman** [comment]
Cool. The CLI will let you add as few or as many SVG icons as you want in a folder. It also supports many folders to add different icon sets or styles. It fully supports class names, so you can use CSS to style the icon.

It automates most of the hassle of setting up SVG sprites.

**Reply by kentcdodds** [reply]
That sounds great. Would you have the time and inclination to create [an example](https://github.com/epicweb-dev/epic-stack/blob/main/docs/examples.md) to demonstrate how you recommend integrating them? If I'm happy with that then we can talk about a PR. But examples are always useful and welcome and go a long way to helping me visualize what it would take to get things integrated.

**Reply by kiliman** [reply]
I've got an example using SVG sprites. I took one of the TailwindUI layouts that use `heroicons` and replaced it with sprites. I'm working on the documentation and will send you the link when it's published.

<img width="1066" alt="image" src="https://github.com/epicweb-dev/epic-stack/assets/47168/6245c68f-9357-4692-ac11-adf4a1aa964d">

**Reply by kentcdodds** [reply]
Might I suggest you not make many changes and instead make minimal changes so it's obvious to folks what changes they need to make. You could just add SVGs to the menu or the theme switcher. The examples documentation explains how to optimize your example so folks can easily tell what changes are necessary.

**Reply by kiliman** [reply]
Ok, I updated the base template.

<img width="237" alt="image" src="https://github.com/epicweb-dev/epic-stack/assets/47168/940163d9-665f-4d1c-947a-12174a57a8dd">

<img width="121" alt="image" src="https://github.com/epicweb-dev/epic-stack/assets/47168/3defa740-6dd3-4a34-96c4-958404e6c865">

<img width="113" alt="image" src="https://github.com/epicweb-dev/epic-stack/assets/47168/40f0a8ee-05cc-446b-a5a7-50eeeab56bec">

**Reply by kiliman** [reply]
Ok, I posted the example with step-by-step instructions as well as individual commits.

https://github.com/kiliman/epic-stack-with-svg-sprites

**Reply by kentcdodds** [reply]
This looks great! Thank you so much for putting this together. I'm happy to merge a PR to examples if you'd like to open one. But if you'd like to get it built in that may not be necessary. Before it's built in though, I'd prefer to make a couple changes. Ultimately, the biggest change I'd like is for the generated code to be this instead:

```tsx
import { type SVGProps } from "react";
import href from "./sprite.svg";
export { href };

export type IconName = 'arrow-left-on-rectangle' | 'pencil-square' | 'user'

function Icon({ icon, ...props}: SVGProps<SVGSVGElement> & { icon: IconName }) {
  return (
    <svg {...props}>
      <use href={`${href}#${id}`} />
    </svg>
  );
}
```

That way you have less code generated, you can pass an `id` to the SVG, and you don't have to import a bunch of individual components if you're using a bunch of icons (or use a namespace specifier).

One thing I wasn't certain of is why the `className` prop was passed to both the `svg` element and the `use` element.

Thanks for taking the time to work on this. It's really useful!

**Reply by kiliman** [reply]
I actually had it where it was a generic `Icon` component with typed ids. The problem is when you want to pass a reference to an icon, like if you have an array of items for your navbar that has a title and icon. You'd have to create an instance to pass it, so that also means you need to include the classes, etc.

```ts
const navigation = [
  { name: 'Dashboard', href: '#', icon: HomeIcon, current: true },
  ...
]
vs 
const navigation = [
  { name: 'Dashboard', href: '#', icon: <Icon id="home" className="h-6 w-6 text-foreground" />, current: true },
  ...
]
```

As for the `className` prop, when I was first testing it, it didn't appear to cascade the class to the `<use>`. I just checked, and it doesn't appear to need the class now. I made some changes to how the sprite file was generated, so that probably fixed my previous issue.

Anyway, if you prefer the "id" method, then I can make that an option from the generator. I like the individual exports. It matches how most icon libraries already work, so it would be a drop-in replacement. I verified that it only bundles the icon components you use. I imported 3 icons from a 200+ icon set.

**Reply by kentcdodds** [reply]
I'm not at all concerned about the array issue you mentioned. I don't expect that to be a common problem and if it is, you could simply pass the icon name instead of the icon element itself 🤷‍♂️

Fixing the `className` would be great 👍

Having an option for the generator would be great. I definitely prefer the single component with typed `icon` prop (don't use `id` because sometimes I may need to pass an `id` to the svg element).

**Reply by kiliman** [reply]
Ok, this is what I've come up with:
```ts
import { type SVGProps } from "react";
import href from "./sprite.svg";
export { href };

export default function Icon({ icon, ...props}: SVGProps<SVGSVGElement> & { icon: IconName }) {
  return (
    <svg {...props}>
      <use href={`${href}#${icon}`} />
    </svg>
  );
}

type IconName =
  | "bookmark"
  | "envelope-open"
```

Then if you add the `--components` flag to the generator, it will also created named components for each icon:
```ts
export function BookmarkIcon(props: SVGProps<SVGSVGElement>) {
  return <Icon icon="bookmark" {...props} />;
}

export function EnvelopeOpenIcon(props: SVGProps<SVGSVGElement>) {
  return <Icon icon="envelope-open" {...props} />;
}
```
You'd use it like this:
```ts
import {
  default as RadixIcon,
  href as radixIcons,
} from "~/components/radixicons";

<RadixIcon icon="bookmark" className="text-red-500 h-6 w-6" />
<RadixIcon icon="envelope-open" className="text-green-500 h-6 w-6" />
```

**Reply by kentcdodds** [reply]
Yup, that's great 👍 Is that working currently? I'm happy to follow your instructions to make it built-in if so 👍

**Reply by kiliman** [reply]
Simplified to use arrow functions
```ts
export const BookmarkIcon = (props: SVGProps<SVGSVGElement>) => <Icon icon="bookmark" {...props} />;
export const EnvelopeOpenIcon = (props: SVGProps<SVGSVGElement>) => <Icon icon="envelope-open" {...props} />;
```

**Reply by kiliman** [reply]
I've updated both the CLI and the example repo. Let me know if you have any questions.

**Comment by kentcdodds** [comment]
This has been implemented #229
